### DISCUSSION TITLE
my own quantized mesh as terrain

### DISCUSSION THREAD

**Discussion by imakihi** [TerriaJS/terriajs#5006] [Q&A] [answered]
Hi there. This is my first question in this discussion. We just started using Terria.io and I have one question regarding terrain data.

So, we have a quantized mesh dataset with layer.json. I tried to use our own terrain data with following statement in my terria.json file.

`			{
                "name": "test",
                "type": "composite",
                "items": [
                    {
                        "name": "Japanese standard tile map",
                        "type": "url-template",
                        "url": "https://cyberjapandata.gsi.go.jp/xyz/std/{z}/{x}/{y}.png",
                        "minimumLevel": 1,
                        "maximumLevel": 18
                    },
                    {
                      "name": "Test Terrain",
                      "type": "cesium-terrain",
                      "url": "data/terrain2/terrain/layer.json"
                    }
                ]
            },
`
However, I haven't had any luck to show my terrain. Can someone gives me some suggestions how I can show my terrain data correctly? I researched and read following discussion but I still couldn't figure out a way to show my terrain...

https://gitter.im/TerriaJS/terriajs?at=5d06f79ea8d9871b329e7b20

Please help me!

Thank you.

Hiroo

**Comment by hardreddata** [comment]
I think you need to ensure your data are in Cesium terrain format per https://cesium.com/docs/cesiumjs-ref-doc/CesiumTerrainProvider.html

You may be able to upload a more generic data set like a raw DEM to Cesium Ion for this conversion.

**Reply by imakihi** [reply]
Hi there. Thank you for your response!
So, I could use my terrain data with Cesium.js without any issues. So, I thought data were okay....
I have no idea but I will try a few more things...
Thanks.

hiroo

**Reply by zoran995** [reply]
@imakihi can you try the following (URL without `layer.json`, cesium uses it like that) You can check the layer `vrTheWorldProvider` in the sandcastle example https://sandcastle.cesium.com/?src=Terrain.html
```
{
  "catalog": [
    { "name": "Test Terrain", "type": "cesium-terrain", "url": "data/terrain2/terrain/" }
  ]
}
```
Just tested with another server I have access to and it works

**Reply by imakihi** [reply]
Thank you for your suggestion, @zoran995. I may be misunderstood on this topic. It sounds like I need to have one of the terrain server to serve my terrain data instead of just pointing my terrain data location on my S3 bucket. Is my understanding correct? 
Thank you for your help!!

**Reply by zoran995** [reply]
I did a test with a national terrain using Nginx to serve a directory with terrain files/directories, just set an URL to root directory where `layer.json` is, not to the file itself. Haven't worked with s3 in a long time, so can't help with that

**Comment by KeyboardSounds** [comment, marked as answer]
To add to @RussellGrew's answer, your data needs to be in one of the terrain formats supported by a TerrainProvider class Cesium. Have a look at the list on the left-hand side [here](https://cesium.com/docs/cesiumjs-ref-doc/CesiumTerrainProvider.html?classFilter=terrainprovider). Cesium terrain is one of them, as well as ArcGISTiledElevation and some others. If your format isn't supported, you could potentially implement your own TerrianProvider.

**Comment by zoran995** [comment]
@KeyboardSounds Is there a description of propper way to make Terria use our own terrain files on `next` branch instead of CesiumIon?
Also, I am wondering how Cesium will play with terrain files for the smaller area (country or city), the rest of the world will be considered as an `EllipsoidTerrain` or something additional is needed?

**Reply by KeyboardSounds** [reply]
As far as I know, using custom terrain for a map shouldn't be too different to using a different basemap, but I don't have any examples for you, sorry :(.
As for Cesium Ion using the ellipsoid outside the area that you have terrain for, yes, that's what happens. Have a look at the NSW DEM [here](https://nsw.digitaltwin.terria.io/#share=s-aOJ2sJHiUi1EF6Q0czxK1pHlW75). Outside NSW (go north a long way to the top of Australia) you can see that the terrain is smooth.

**Reply by hardreddata** [reply]
As an aside, this ticket makes me chuckle

https://github.com/CesiumGS/cesium-ion-rest-api-examples/issues/294

A good reason to want to explore hosting ones own terrain.

**Reply by zoran995** [reply]
looks like I run into the Cesium issue with requiring the global terrain, https://github.com/CesiumGS/cesium/issues/6598. Will go with kring advice on no height tile, or try to work out Cesium to support composite terrain (combine multiple terrain sources in one)
